test/hotspot/jtreg/serviceability/jvmti/events/FieldAccess/FieldsEventsFromJNI/libJvmtiFieldEventsFromJNI.cpp

```diff
@@ -0,0 +1,198 @@
+    fatal(jni, "The access field is incorrect.");
+  }
+  deallocate(jvmti,jni, f_name);
+
+  char* obj_class_name = get_object_class_name(jvmti, jni, object);
+  LOG("The object class '%s'\n", obj_class_name);
+  if (strcmp(obj_class_name, "LFieldsEventsFromJNI;") != 0) {
+    fatal(jni, "The fields's class name is incorrect.");
+  }
+  deallocate(jvmti,jni, obj_class_name);
+
+  modify_cnt++;
+}
+
+JNIEXPORT jint JNICALL
+Agent_OnLoad(JavaVM *vm, char *options, void *reserved) {
+  jvmtiEnv *jvmti = nullptr;
+  jint res = vm->GetEnv((void **) &jvmti, JVMTI_VERSION_21);
+  if (res != JNI_OK) {
+    return JNI_ERR;
+  }
+  jvmtiError err = JVMTI_ERROR_NONE;
+  jvmtiCapabilities capabilities;
+  (void) memset(&capabilities, 0, sizeof (capabilities));
+  capabilities.can_generate_field_access_events = true;
+  capabilities.can_generate_field_modification_events = true;
+  err = jvmti->AddCapabilities(&capabilities);
+  check_jvmti_error(err, "AddCapabilities");
+  jvmtiEventCallbacks callbacks;
+  (void) memset(&callbacks, 0, sizeof (callbacks));
+  callbacks.FieldAccess = &cbFieldAccess;
+  callbacks.FieldModification = &cbFieldModification;
+  err = jvmti->SetEventCallbacks(&callbacks, (int) sizeof (jvmtiEventCallbacks));
+  check_jvmti_error(err, "SetEventCallbacks");
+  jvmti_env = jvmti;
+  return JNI_OK;
+}
+
+extern "C" {
+JNIEXPORT void JNICALL
+Java_FieldsEventsFromJNI_enableEventsAndAccessField(
+    JNIEnv *jni, jobject self, jboolean isEventExpected, jthread eventThread) {
+
+  jvmtiError err = JVMTI_ERROR_NONE;
+
+  jclass cls = jni->GetObjectClass(self);
+  if (cls == nullptr) {
+    fatal(jni, "No class found");
+  }
+  jfieldID fieldToRead = jni->GetFieldID(cls, "accessField", "Ljava/lang/String;");
+  if (fieldToRead == nullptr) {
+    fatal(jni, "No field found");
+  }
+
+  // Set watch and access field without returning to calling java code
+  access_cnt = 0;
+  err = jvmti_env->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_FIELD_ACCESS, eventThread);
+  check_jvmti_error(err, "SetEventNotificationMode");
+  err = jvmti_env->SetFieldAccessWatch(cls, fieldToRead);
+  check_jvmti_error(err, "SetFieldAccessWatch");
+
+  jstring jname = (jstring)jni->GetObjectField(self, fieldToRead);
+
+  err = jvmti_env->ClearFieldAccessWatch(cls, fieldToRead);
+  check_jvmti_error(err, "ClearFieldAccessWatch");
+  err = jvmti_env->SetEventNotificationMode(JVMTI_DISABLE, JVMTI_EVENT_FIELD_ACCESS, eventThread);
+  check_jvmti_error(err, "SetEventNotificationMode");
+
+  const char* name_str = jni->GetStringUTFChars(jname, nullptr);
+  LOG("The field %s\n", name_str);
+  if (strcmp(name_str, "accessFieldValue") != 0) {
+    fatal(jni, "The field value is incorrect.");
+  }
+
+  if (access_cnt != isEventExpected) {
+    fatal(jni, "The field access count is incorrect.");
+  }
+  jni->ReleaseStringUTFChars(jname, name_str);
+}
+
+JNIEXPORT void JNICALL
+Java_FieldsEventsFromJNI_enableEventsAndModifyField(
```

Comment:
Need to rename these as well: `s/Java_FieldsEventsFromJNI/Java_FieldEventsFromJNI/g`.